proposals/0000-fine-grained-unused-warnings.rst

```diff
@@ -0,0 +1,394 @@
+
+Looking at this, it might appear that ``bar`` is indirectly unused, and will thus produce a warning under this proposal. However, this is not the case: ``foo`` does not qualify as directly unused, since its relevant warning flag ``-Wunused-top-binds`` is not active, and so ``bar`` also does not qualify as indirectly unused. Thus, there will still not be any warnings produced by the code under this proposal.
+
+
+Effect and Interactions
+-----------------------
+For the most part, the effects of this proposal are minor. The main differences for existing
+code-bases are that the warning message GHC prints for the indirect warnings
+is different, although due to the more consistent treatment of warning flags, existing code bases can also sometimes
+get more or fewer warnings in cases of indirectly unused bindings where two warning flags interact with one another. There can also be additional warnings about indirectly unused imports and `forall` binds.
+
+Since the warnings don't have any special formats, existing tools should be able to handle them without issues.
+
+Users that don't wish to see warnings about indirectly unused bindings can turn those warnings off.
+
+Costs and Drawbacks
+-------------------
+The warning mechanism is somewhat more complicated and as a consequence might
+have a somewhat higher maintenance cost, which might especially concern listing out all the other bindings that an indirectly unused binding is referenced by.
+
+Alternatives
+------------
+* We could combine warnings of unused bindings and the indirect non-uses they induce. This could be similar to how error locations are combined in a single error message for duplicate declarations.
+
+  * A possible advantage is that we could simplify the mechanism by removing the configurability of turning the warnings off.
+    This would still give us the benefit of reducing the potential for confusion from these warnings, however users that wish to turn these warnings off could not do so.
+
+  * A disadvantage is that most third-party tools dealing with error messages will likely have a harder time parsing the warning messages.
+
+* A different name could be chosen for the new flag, ``-Windirectly-unused-binds``. For example:
+  * ``-freport-indirect-uses``
+
+* Instead of ``-Windirectly-unused-binds``, we could separate each warning flag (like ``-Wunused-imports``)
+  into two (like ``-Windirectly-unused-imports`` and ``-Wdirectly-unused-imports``) and a warnings group like ``-Wno-indirect-uses`` to turn off all warnings about indirectly unused bindings at once.
+
+  * This would offer more configurability if users want to see some warnings about indirectly unused bindings but not others.
+
+  * It would require a higher number of warning flags.
+
+Implementation Plan
+-------------------
+
+`@Jadefalke256 <https://github.com/Jadefalke256>`_ has `expressed interest <https://gitlab.haskell.org/ghc/ghc/-/issues/20190#note_505317>`_ in implementing this proposal.
```

Comment:
The link to the user 404s. I hope they just deleted their github account and are still interested in implementing this?

Reply:
Not sure, they might have renamed their account. I was planning on asking them again once the proposal is accepted. If they don't have time or motivation anymore, I will find time to implement it.